Update the new modify properties API to automatically retry on concurrent modification exception.

keith-turner opened the pull request.

keith-turner commented:
Meant to open this PR as a draft.  Only updated table operations.  Still need to update namespace and instance ops, but wanted to wait on those to see if there as any feedback on table operations.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:
```diff
@@ -1035,6 +1031,30 @@ public void modifyProperties(String tableName, final Consumer<Map<String,String>
   }
 
+  @Override
+  public void modifyProperties(String tableName, final Consumer<Map<String,String>> mapMutator)
```
Does it make sense to return a Future<Void> here instead? That would allow the caller to potentially perform other logic before checking the result and the caller has to handle the exceptions when calling Future.get().

keith-turner replied:
I would like to see Accumulo support async APIs.  I think that should be done with consideration of the entire API and looking at what APIs we want to add.  Not sure about doing that as a one off here.  Would need to create and manage a thread  or thread pool to do that here, which is another reason to consider it API wide maybe (maybe we want to allow passing in an executor for async APIs?)

I would also like to explore making the read and write pipelines in the tserver and sserver asynchronous.  I think this could support more client connections per server and would be good for making them  more responsive in the case we had discussed in the past of having many sservers and few tservers.  It think it would be good to make the server pipelines async at the same time as adding async APIs to Accumulo so that the two could influence each other.

keith-turner replied:
Async APIs and server pipelines may be a good 3.0 project.

cshannon replied:
> Async APIs and server pipelines may be a good 3.0 project.

I agree with this, I think adding Async APIs for the next major version would be a nice addition.

EdColeman commented on core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:
```diff
@@ -1035,6 +1031,30 @@ public void modifyProperties(String tableName, final Consumer<Map<String,String>
+    Retry retry =
+        Retry.builder().infiniteRetries().retryAfter(25, MILLISECONDS).incrementBy(25, MILLISECONDS)
+            .maxWait(30, SECONDS).backOffFactor(1.5).logInterval(3, MINUTES).createRetry();
+
```
Is it correct to have the log interval greater than max wait? Does this efficiently eliminate logging except on timeout?

keith-turner replied:
Good question.  I copied and pasted that code and modified it a bit. I went a looked at the code and logging does not happen unless an explicit method call is made.  I need to look into this some more, see if I should call the logging method or remove the config from the builder.

keith-turner replied:
I looked more closely at the retry code and made some updates in 3873e94.

Below are some log messages I saw while running the new IT added in this PR.

```
2022-09-28T14:31:47,217 [admin.TableOperations] DEBUG: Unable to modify table properties for PropStoreConfigIT_concurrentTablePropsModificationTest0 because of concurrent modification, retrying attempt 1 (suppressing retry messages for 180000ms)
2022-09-28T14:31:47,217 [util.Retry] DEBUG: Sleeping for 25ms before retrying operation
2022-09-28T14:31:47,252 [admin.TableOperations] DEBUG: Unable to modify table properties for PropStoreConfigIT_concurrentTablePropsModificationTest0 because of concurrent modification, retrying attempt 1 (suppressing retry messages for 180000ms)
2022-09-28T14:31:47,252 [util.Retry] DEBUG: Sleeping for 25ms before retrying operation
2022-09-28T14:31:47,283 [util.Retry] DEBUG: Sleeping for 62ms before retrying operation
2022-09-28T14:31:47,330 [util.Retry] DEBUG: Sleeping for 150ms before retrying operation
2022-09-28T14:31:47,332 [util.Retry] DEBUG: Sleeping for 158ms before retrying operation
2022-09-28T14:31:47,362 [util.Retry] DEBUG: Sleeping for 79ms before retrying operation
2022-09-28T14:31:47,458 [util.Retry] DEBUG: Sleeping for 111ms before retrying operation
2022-09-28T14:31:47,497 [util.Retry] DEBUG: Sleeping for 217ms before retrying operation
2022-09-28T14:31:47,500 [util.Retry] DEBUG: Sleeping for 215ms before retrying operation
2022-09-28T14:31:47,578 [util.Retry] DEBUG: Sleeping for 153ms before retrying operation
2022-09-28T14:31:47,727 [util.Retry] DEBUG: Sleeping for 302ms before retrying operation
2022-09-28T14:31:47,734 [admin.TableOperations] DEBUG: Unable to modify table properties for PropStoreConfigIT_concurrentTablePropsModificationTest0 because of concurrent modification, retrying attempt 1 (suppressing retry messages for 180000ms)
2022-09-28T14:31:47,735 [util.Retry] DEBUG: Sleeping for 25ms before retrying operation
2022-09-28T14:31:47,746 [admin.TableOperations] DEBUG: Unable to modify table properties for PropStoreConfigIT_concurrentTablePropsModificationTest0 because of concurrent modification, retrying attempt 1 (suppressing retry messages for 180000ms)
2022-09-28T14:31:47,746 [util.Retry] DEBUG: Sleeping for 25ms before retrying operation
```

keith-turner commented:
> As a follow-on - would it make sense to move Retry from fate to core.util package? Its being used outside of fate now (here and elsewhere)

That sounds reasonable.  Also I think there is some pre-existing code in accumulo that does retries w/o using the Retry class, code that existed before Retry was created.

cshannon reviewed (commented):
This approach seems reasonable to me for handling the ConcurrentModificationException as would be a better experience for the user to not have the exception bubble up. I also agree with @EdColeman that the Retry utility could be moved outside of Fate as more of a general use utility but that could be done in a different PR if not in this one to keep the scope smaller.

ctubbsii reviewed (changes requested):
I would prefer we not automatically retry. I would rather throw the ConcurrentModificationException. This should happen rarely, but when it does, it should be made known to the user, in case they want to do something different... like check with other groups who made the other change to see what the actually configuration *should* be. And, we shouldn't be hiding potential errors in user's automation, if they unintentionally have multiple processes changing the configuration without being aware of that.

I think for maximum flexibility, and minimal code maintenance burden on our part, it would be better to let users decide how to handle retries if they wish.

So, I don't think we should do this.

EdColeman commented:
@ctubbsii - it the exception is going to be thrown - would it be better to have it as a checked exception to force handling or a deliberate choice to pass up the stack? (ConcurrentModificationException inherits from RuntimeException)

keith-turner commented:
> I would prefer we not automatically retry. I would rather throw the ConcurrentModificationException. This should happen rarely, but when it does, it should be made known to the user, in case they want to do something different... like check with other groups who made the other change to see what the actually configuration should be. And, we shouldn't be hiding potential errors in user's automation, if they unintentionally have multiple processes changing the configuration without being aware of that.

For the case you describe no one could ever rely on the concurrent modification exception to ensure the correctness of their system.  If they want to ensure only a single process modifies configuration this exception does not offer a reliable way to ensure that or detect when that is not the case.  I see the exception as a nuisance that may occur for normal use cases in a distributed system, it does not offer anything of value and will just cause people problems IMO.

The wonderful thing about this new API is that allows a user the chance to atomically inspect the config and modify it.  This new ability to atomically inspect and modify allows user to build code that is correct in a distributed system on top of this API.  If someone wants to utilize this inspect and modify capability I can only think of uses cases where they would want to automatically retry.  If their mapMutator inspects before modfifying then it will automatically redo the inspection on the latest snapshot when there is a concurrent modification.  Atomic inspect and modify is a really neat new feature and I think automatic retry makes it easier to use it correctly.

keith-turner commented:
The discussion of retry vs not retry made me realize the javadoc could be improved to provide more detail about this new feature.  I update some of the javadoc in [7850b0c](https://github.com/apache/accumulo/pull/2967/commits/7850b0ccea80c635c921e643fd712d20895b3de9)

EdColeman commented on core/src/main/java/org/apache/accumulo/core/clientImpl/NamespaceOperationsImpl.java:
```diff
@@ -20,20 +20,11 @@
-import static org.apache.accumulo.core.util.Validators.EXISTING_NAMESPACE_NAME;
-import static org.apache.accumulo.core.util.Validators.NEW_NAMESPACE_NAME;
+import static java.util.concurrent.TimeUnit.*;
+import static org.apache.accumulo.core.util.Validators.*;
```
Star imports will cause the QA check to fail

EdColeman commented on core/src/main/java/org/apache/accumulo/core/clientImpl/NamespaceOperationsImpl.java:
```diff
@@ -227,6 +216,34 @@ public void modifyProperties(final String namespace,
+        break;
+      } catch (ConcurrentModificationException cme) {
+        try {
+          retry.logRetry(log, "Unable to modify namespace properties for " + namespace
```
Maybe it could read - "modification attempt failed because of a concurrent modification - will retry"

dlmarion replied:
Agree that we should add the fact the fact that we are going to retry here and in other places where Retry.logRetry is called.

keith-turner replied:
I made a top level comment about this.

dlmarion replied:
I think this can be resolved based on the top level comment.

EdColeman commented on test/src/main/java/org/apache/accumulo/test/conf/PropStoreConfigIT.java:
```diff
@@ -27,11 +27,17 @@
 import java.util.function.Supplier;
 
-import org.apache.accumulo.core.client.Accumulo;
+import org.apache.accumulo.core.client.*;
```
another star import

EdColeman reviewed (commented):
The code looks good overall - some minor suggestions on messages so they indicate that the concurrent modification error is transitory and  the update will be retried.

There are also star imports in some of the files - this will fail the QA checks.

I keep going back and forth on automatically retrying - there are benefits to both approaches.  Signaling to a user that a concurrent modification occurred and their update failed seems a reasonable approach.  On the other hand, when a change is submitted, nothing prevents another change coming along right behind and changing things to another value. Setting a property is not a guarantee that it would not be changed when it is used - so this change would not impact that.

keith-turner commented:
> On the other hand, when a change is submitted, nothing prevents another change coming along right behind and changing things to another value. Setting a property is not a guarantee that it would not be changed when it is used - so this change would not impact that.

Yeah that is how I see it.  If there is a good change followed by a problematic change, the time between those could be zero,3 seconds, or 4 months.  Throwing a concurrent modification exception offers no reliable solution to this.  Using conditional updates does offer a possible solution if all updates use them because some problematic changes may not be made.

ctubbsii commented:
> For the case you describe no one could ever rely on the concurrent modification exception to ensure the correctness of their system. If they want to ensure only a single process modifies configuration this exception does not offer a reliable way to ensure that or detect when that is not the case. I see the exception as a nuisance that may occur for normal use cases in a distributed system, it does not offer anything of value and will just cause people problems IMO.

What I'm suggesting is that internally we check the version and ensure it hasn't changed. If the version has changed because somebody else updated the ZK node with a different edit, then we throw ConcurrentModificationException. That seems like a pretty reliable way to ensure that other changes are detected.

> The wonderful thing about this new API is that allows a user the chance to atomically inspect the config and modify it. This new ability to atomically inspect and modify allows user to build code that is correct in a distributed system on top of this API. If someone wants to utilize this inspect and modify capability I can only think of uses cases where they would want to automatically retry. If their mapMutator inspects before modfifying then it will automatically redo the inspection on the latest snapshot when there is a concurrent modification. Atomic inspect and modify is a really powerful new feature and I think automatic retry makes it easier to use it correctly.

I agree atomic inspect and modify is very powerful, but one of its powers is the ability to detect collisions. I also agree that this empowers correctly written mapMutator code to ensure the desired outcome, even in the face of collisions.

However, my use case that I'm concerned about isn't about the desired outcome of the map mutation. The aspect I'm concerned about is more about administrators / operators being able to detect rogue threads that are causing collisions in a system that an administrator may wish to detect, track down, and address, before they can trust the map mutator code they are trying to run. It's not a "computer science" concern about correctness... it's a practical "IT" concern about system administration. I think you're overlooking that practical concern, even though I agree with you on the computer science/algorithm correctness aspect and the convenience of retrying.

I don't think this will cause people problems if we document it properly in the Javadoc.

All that said, retry may be okay if there's a warning logged (on both the client and the server side, ideally) that notes the change was detected and it will be retried. However, I still have concerns that it may never finish (if we retry indefinitely; very unlikely, and my concern is proportionately low for this one), and the general restriction on users (my concern is higher for this one). Just because we can't think of a use case where a retry is not desired, it doesn't mean there isn't one, and I see no reason to restrict users by adding extra code we have to maintain, just because we failed to have the foresight to think of a way a use case a user might have.

ctubbsii commented:
> > On the other hand, when a change is submitted, nothing prevents another change coming along right behind and changing things to another value. Setting a property is not a guarantee that it would not be changed when it is used - so this change would not impact that.
> 
> Yeah that is how I see it. If there is a good change followed by a problematic change, the time between those could be zero,3 seconds, or 4 months. Throwing a concurrent modification exception offers no reliable solution to this. Using conditional updates does offer a possible solution if all updates use them because some problematic changes may not be made.

I agree with that. I'm not necessarily suggesting we need to try to detect that. I'm more concerned that we're adding extra code that goes out of our way to pretend it didn't happen and hide the fact from the user when we did detect it. I don't see a reason to do that. I don't think we should try to detect the case where it could happen silently... I'm just saying we shouldn't be the ones perpetuating the silence when it could have otherwise been made known to the user that it did happen.

keith-turner commented:
> All that said, retry may be okay if there's a warning logged (on both the client and the server side, ideally) that notes the change was detected and it will be retried. 

This PR is using the Retry class and logging will occur via it when retries happen.

ctubbsii commented:
> > All that said, retry may be okay if there's a warning logged (on both the client and the server side, ideally) that notes the change was detected and it will be retried.
> 
> This PR is using the Retry class and logging will occur via it when retries happen.

Is the intent to retry indefinitely, or are we going to decide on a number of retries for the user, or are we going to add additional complexity to allow them to tune the retry, and customize the other retry options, like retry delay or back-off? How deep did you want to go with this retry?

cshannon commented:
Retry forever would make me a little concerned. Even if  _should_ be ok in this case because it should eventually succeed at some point weird and unexpected things can sometimes happen (especially in distributed systems). I would think at some point it would be good to terminate and throw an exception as a fail safe but I don't know at what point to do that and whether or is worth the complexity to do so and make it configurable or just pick an arbitrary amount of times to try.

dlmarion commented:
I had asked about returning Future<Void> for this method. Not only to make it async, but cancel-able.

dlmarion commented:
I was thinking of something like:

```
 public Future<Void> modifyProperties(final Consumer<Map<String,String>> mapMutator)
     throws AccumuloException, AccumuloSecurityException, IllegalArgumentException {

   final Retry retry =
       Retry.builder().infiniteRetries().retryAfter(25, MILLISECONDS).incrementBy(25, MILLISECONDS)
       .maxWait(30, SECONDS).backOffFactor(1.5).logInterval(3, MINUTES).createRetry();

   FutureTask<Void> f = new FutureTask<>(() -> {
     while (true) {
       try {
         tryToModifyProperties(mapMutator);
         break;
       } catch (ConcurrentModificationException cme) {
         try {
           retry.logRetry(LoggerFactory.getLogger(InstanceOperationsImpl.class),
               "Unable to modify instance properties for because of concurrent modification");
           retry.waitForNextAttempt();
         } catch (InterruptedException e) {
           throw new RuntimeException(e);
         }
       } finally {
         retry.useRetry();
       }
     }
   }, null);
   
   ForkJoinPool.commonPool().execute(f);
   
   return f;
 }
```

Whether or not the ForkJoinPool is used or not doesn't really matter to me. Other places in the client code where `ClientContext.threadPools()` is used an ExecutorService is created and then shutdown in the same method. I used ForkJoinPool here for convenience as we don't need a pool of Threads.

ctubbsii commented:
> I was thinking of something like:

I think it might be useful to have the Future return something useful, like number of retries that occurred before success, or how long it took to succeed. Okay, maybe not those... but maybe there's something else useful? Also, I'm wondering if we can override the cancel method on the future to specifically cancel the retries by setting a flag and breaking out of the retry loop, rather than relying on interrupting the thread and dealing with exception handling when we cancel.

cshannon commented:
> > I was thinking of something like:
> 
> I think it might be useful to have the Future return something useful, like number of retries that occurred before success, or how long it took to succeed. Okay, maybe not those... but maybe there's something else useful? Also, I'm wondering if we can override the cancel method on the future to specifically cancel the retries by setting a flag and breaking out of the retry loop, rather than relying on interrupting the thread and dealing with exception handling when we cancel.

I'm not sure I like the idea of overriding the cancel method. Handling the InterruptedException is the standard way of handling a canceled future so I don't think we should deviate from that pattern as that's the intended way for handling it.

In terms of returning something useful the main thing the user cares about is it succeeded so I don't see a problem just returning nothing or true. I don't think returning number of retrys or anything like that would provide a lot of value. Maybe you could return a map of the properties after update which would be the current state of all the properties at the scope (instance, namespace, table, etc)? The user would already know the specific properties that were changed and their values as they were supplied in the mapMutator so I don't think it makes sense to return that but returning the state of all properties might be nice. Of course the user could also just call the get properties method if they really wanted to see that.

ctubbsii commented:
> I'm not sure I like the idea of overriding the cancel method. Handling the InterruptedException is the standard way of handling a canceled future so I don't think we should deviate from that pattern as that's the intended way for handling it.

Okay, that's fine. I haven't looked into it very closely. I was just thinking it might be nice to have some degree of confidence on whether it succeeded or not upon cancellation, and we would know that if we interrupted between retries.

keith-turner commented:
@ctubbsii @cshannon @dlmarion  I did three experiments in [406b9f8](https://github.com/apache/accumulo/pull/2967/commits/406b9f84e96faf936e071797e125f11b8118d00b) where I tried different ways of limiting the retries and returning the final map of props.  Did this based on reading to the feedback from yesterday just to see how different approaches looked.  I did the following :

 * I made instanceOperations().modifyProperties() take a timeout duration.
 * I renamed namespaceOperations().modifyProperties() to modifyPropertiesAsync() and it returns a CompletableFuture.  It creates a cached thread pool (can't shut the thread pool down before method exits so the cached thread pool is good option) to execute the task asynchronously.  I was wary of using the common pool from fork join because it shared JVM wide and is intended for computational task and not I/O task.  I added Async to the name to make it clear that the operation may not be complete when the method returns.
 * I made tableOperations().modifyProperties() return a CompleteableFuture and take an Executor as parameter that it uses to execute the task.  This avoids the Accumulo code having to create and manager an executor to run the async task.

I made all of the operations return a `Map<String, String>`.  The returned map is the final map that was accepted on the last retry.  I think this will be useful for conditional updates as it allows the user to inspect what the final result was w/o having to do anything odd in their lambda like setting an atomic reference. 

To get a feel for how all of these changes would look when used I added three new test that make two conditional updates using each API.   These three test were added at the end of PropStoreConfigIT.java.

After making all these change I was digging a bit deeper in the code and found that the current code that is committed will already retry forever.  The currently committed code calls [executeVoidTableCommand()](https://github.com/apache/accumulo/blob/8a7ed07d74a0ba0d127a37a5fced23d34a270dfe/core/src/main/java/org/apache/accumulo/core/rpc/clients/ManagerThriftClient.java#L101-L136) which retries forever when there are problems communicating with the Manager.  After seeing this I think retrying forever is a good option in this PR for the following reasons :

 * The code to communicate with the manager retries forever and many table operations use this.
 * When things are working well (Manager and Zookeeper are healthy) I think its extremely unlikely that any modifyProperties call would retry forever given they are doing exponential backoff.  When that code does have a high chance of retrying forever is when ZK and/or manager are unhealthy and in some code already retries forever in those cases.

It would still be nice to return the final Map that was accepted.

ctubbsii commented:
Thoughts on those experiments (skip over for my conclusion):

* I'm not sure it's necessary to name the method "Async()" when it returns a Future. I think that's implied. But, I suppose it doesn't hurt to be explicit. I just wouldn't want to have two versions.... one async and one synchronous, because we imply it with the name suffix, because that name suffix is often used to distinguish between two versions of the same command in many other APIs (ZooKeeper does this, for one example).
* I agree returning the final map is nice (very convenient)
* I'm not sure the existing retry forever is the same kind of indefinite retry as we've been discussing. That retries trying to communicate with the server. This retry is a new retry when the server communicates fine, but communicates the detection of a collision. I don't think they're comparable.
* I don't think it's a problem to use the ForkJoinPool as a default, but I think it's a bit of API bloat to start accepting executors on all async APIs. I think if we start adding async APIs, we should default to the ForkJoinPool, and provide some other mechanism to set an executor at the time the AccumuloClient is constructed, for use with that client, and shut down when that client is closed. This goes back to previous discussions about having per-client resources, and possibly making those resources sharable across clients only in explicit ways at the time the client is constructed. There shouldn't be any global JVM resources being auto-magically created and used and cleaned up (think SingletonManager stuffs).
* You started using the term "conditional updates" in the method names of the test code. While this method can certainly be used for conditional updates, and the tests themselves may be doing conditional updating, I just want to be clear that this feature itself is not necessarily a conditional update API, because we do not impose any conditional gating. Any conditionals would be imposed by the user in their own mutator code. I mention this only because I think we should be careful not to use the term "conditional" as a general term for this feature. I don't think you were doing that... but it's something that might easily be done if we're not careful, because we have ConditionalMutation, which is very similar, for user data. So, if we're not careful, it could get confusing for users.

My conclusion: Based on the experimentation you've done, and the state of this discussion, and thinking about how much work is still needed to design asynchronous APIs in a way that's consistent going forward, instead of just doing it as a one-off for this specific feature, I'm thinking that I'd prefer to settle with the infinite retry you initially proposed, and to punt on the asynchronous stuff until 3.x, so we can do it consistently.

keith-turner commented:
> I'm not sure it's necessary to name the method "Async()" when it returns a Future. I think that's implied. 

For that I was thinking its in Async method among many sync methods and using the return type is optional.  So wanted to make sure anyone using knew that when the method returned it may not be done.

>  I think if we start adding async APIs, we should default to the ForkJoinPool, and provide some other mechanism to set an executor at the time the AccumuloClient is constructed, for use with that client, and shut down when that client is closed.

I am not sure scoping the executor at the client is good.  Many methods on Completable future have a version that takes an Executor and one that does not.  This allows developer control over limiting how many cores an operation can use.   I think this important if a developer wants to limit one scan to one core and another to 32 cores for example.  If the executor was scoped to the accumulo client, then they would need to create two different clients for the scans.

> My conclusion: Based on the experimentation you've done, and the state of this discussion, and thinking about how much work is still needed to design asynchronous APIs in a way that's consistent going forward, instead of just doing it as a one-off for this specific feature

I agree.  I also mentioned previously it would be good to do the async client API work at the same time as making the servers support async read/write pipelines so that the two could influence each in a development cycle.  This is another reason to defer this to 3.0.

Another little thing that I noticed while experimenting with an async API for this PR is that our existing checked exceptions would be wrapped by something like an ExecutionException when using futures.  Its a very nuanced thing, but exceptions may be something else to consider in the design of async APIs.  I noticed that ComletableFuture introduced a new method called [join()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletableFuture.html#join()) that is similar to [get()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletableFuture.html#get()) except it has no checked exceptions.  I really like it, so much easier to use.  Makes me wonder if there is a wider move away from checked exceptions in the Java APIs because they are painful to use w/ lambdas.

I think I will revert the experimental commmit I did and then do another commit to add the map return type.

ctubbsii commented:
> I think I will revert the experimental commmit I did and then do another commit to add the map return type.

There's conflicts with main, so if you could rebase it onto main, or merge main into this branch to resolve the conflicts, that would also be helpful to make it easier to merge once you're done.

keith-turner commented:
> There's conflicts with main, so if you could rebase it onto main, or merge main into this branch to resolve the conflicts, that would also be helpful to make it easier to merge once you're done.

I merged main in, fixed a bunch of build problems I created, and updated the javadocs.

DomGarguilo commented:
Not sure if this adds any value but I saw an test failure when running `PermissionsIT.systemPermissionsTest` locally on main that looked like it could be related:
```java
2022-10-03T13:08:18,940 [functional.PermissionsIT] DEBUG: Verifying the DROP_TABLE permission
2022-10-03T13:08:19,940 [functional.PermissionsIT] DEBUG: Confirming that the lack of the DROP_TABLE permission properly restricts the user
2022-10-03T13:08:20,329 [functional.PermissionsIT] DEBUG: Confirming that the presence of the DROP_TABLE permission properly permits the user
2022-10-03T13:08:21,000 [functional.PermissionsIT] DEBUG: Verifying the ALTER_TABLE permission
2022-10-03T13:08:22,000 [functional.PermissionsIT] DEBUG: Confirming that the lack of the ALTER_TABLE permission properly restricts the user

java.util.ConcurrentModificationException
	at org.apache.accumulo.core.rpc.clients.ManagerThriftClient.executeVoidTableCommand(ManagerThriftClient.java:128)
	at org.apache.accumulo.core.rpc.clients.ManagerThriftClient.executeVoid(ManagerThriftClient.java:142)
	at org.apache.accumulo.core.clientImpl.TableOperationsImpl.modifyProperties(TableOperationsImpl.java:1027)
	at org.apache.accumulo.test.functional.PermissionsIT.testMissingSystemPermission(PermissionsIT.java:221)
	at org.apache.accumulo.test.functional.PermissionsIT.systemPermissionsTest(PermissionsIT.java:127)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.api.AssertTimeout.lambda$assertTimeoutPreemptively$4(AssertTimeout.java:138)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: ThriftConcurrentModificationException(description:Failed to modify properties to zooKeeper for TablePropKey{TableId=4'}, properties changed since reading.)
	at org.apache.accumulo.core.manager.thrift.ManagerClientService$modifyTableProperties_result$modifyTableProperties_resultStandardScheme.read(ManagerClientService.java:9258)
	at org.apache.accumulo.core.manager.thrift.ManagerClientService$modifyTableProperties_result$modifyTableProperties_resultStandardScheme.read(ManagerClientService.java:9217)
	at org.apache.accumulo.core.manager.thrift.ManagerClientService$modifyTableProperties_result.read(ManagerClientService.java:9143)
	at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:88)
	at org.apache.accumulo.core.manager.thrift.ManagerClientService$Client.recv_modifyTableProperties(ManagerClientService.java:263)
	at org.apache.accumulo.core.manager.thrift.ManagerClientService$Client.modifyTableProperties(ManagerClientService.java:247)
	at org.apache.accumulo.core.clientImpl.TableOperationsImpl.lambda$modifyProperties$1(TableOperationsImpl.java:1028)
	at org.apache.accumulo.core.rpc.clients.ManagerThriftClient.executeVoidTableCommand(ManagerThriftClient.java:107)
	... 17 more
```
This doesn't happen all the time but I am able to reproduce the failure after a few runs.

dlmarion commented:
I just saw a CME in PropStoreConfigIT.modifyTablePropTest

keith-turner commented:
@dlmarion and @DomGarguilo for the CME errors you mentioned, you are seeing those in the main branch w/o these changes merged?

DomGarguilo commented:
> @dlmarion and @DomGarguilo for the CME errors you mentioned, you are seeing those in the main branch w/o these changes merged?

Yes, main branch. Haven't tried with the changes in this PR yet.

dlmarion commented:
Correct

keith-turner commented:
> Yes, main branch. Haven't tried with the changes in this PR yet. Might have been the wrong spot to report the test failure I saw. Do you think I should open a separate ticket instead?

This is a good place to mention that, seems related.  If this PR does not fix the problem then we could open an issue.

EdColeman commented:
Is this a concurrency issue it the test?  I would help to know where multiple, concurrent change requests being made.  While this pr could "fix" the issue - it is surprising that it is happening at all.

I am assuming that these are new with the adoption of using the new atomic,  multi-property set functions in new places in the code - but those changes should have just been grouping prop changes that were occurring serially into a single call.

It may help too if the underlying exception printed the expected vs the received data versions - that may help unwind the timeline of these events.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:
```diff
@@ -57,19 +56,80 @@ void setProperty(final String property, final String value)
    * property overrides in ZooKeeper. Only properties which can be stored in ZooKeeper will be
    * accepted.
    *
+   * <P>
```
Please use lowercase for html tags to comply with HTML5 conventions and W3C recommendations (and are entirely disallowed in xhtml). I keep fixing these when they get added as uppercase, but they keep getting added using uppercase again, and they are really hard to grep for or use automated tooling to check, because generics also use the less than and greater than symbols, and html is notoriously difficult to parse properly. Parsing inside a javadoc comment is even trickier.

ctubbsii reviewed (dismissed):
Didn't have time to review the full thing, but commenting on the things that stood out right away:

* please use lowercase for HTML5 tags
* the API for the method, modified to return the resulting map looks good.

I didn't look any further.

cshannon commented:
I reviewed the changes and they LGTM. I think it's reasonable to wait and just to do the bigger async update for 3.0. I also do like the return of the final updated map now that retry is supported so the user can see what was accepted. In terms of the CME that has happened a few times during tests, I like @EdColeman's idea of printing out the state to help track down expected vs the actual result. We could also just go ahead and proactively update the logging now to print the state when CME is thrown now as part of this PR.

keith-turner commented:
> Is this a concurrency issue it the test? I would help to know where multiple, concurrent change requests being made. While this pr could "fix" the issue - it is surprising that it is happening at all.

I was also wondering about this.  I thought of another possible cause besides an actual concurrent update.  If the props were read from a stale cache it could also cause this.  So there are two updates that do not overlap in time and the 2nd update reads props from a cache where the first update is not reflected.  Not sure if this can even happen, its been a while since i reviewed the recent server side changes to store props in a single znode.  I remember those changes introduced some new caching.

keith-turner commented:
> We could also just go ahead and proactively update the logging now to print the state when CME is thrown now as part of this PR.

I think I would prefer to get this PR merged now that consensus has been reached on these changes.

@ctubbsii I updated the HTML tags in [b889e95](https://github.com/apache/accumulo/pull/2967/commits/b889e95366070eb48213107aac7c34350426f2cb).  Also I added a paragraph to the javadoc in b889e95 containing some info that I thought was important to call out.  @EdColeman, @cshannon, @ctubbsii, or @dlmarion  could one of you take a look at that new paragraph?

dlmarion commented on core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:
```diff
@@ -56,30 +56,36 @@ void setProperty(final String property, final String value)
    *
-   * <P>
+   * <p>
+   * Accumulo has multiple layers of properties that for many APIs and SPIs presented as single
```
... APIs and SPIs **are** presented as **a** single merged view.

keith-turner replied:
fixed in 767ba52a6192d70ffa8b2152f12127cb7c560c42

dlmarion commented on core/src/main/java/org/apache/accumulo/core/client/admin/NamespaceOperations.java:
```diff
@@ -175,6 +175,12 @@ void setProperty(String namespace, String property, String value)
    * properties but has the same behavior and better documentation.
    *
+   * <p>
+   * Accumulo has multiple layers of properties that for many APIs and SPIs presented as single
```
same comment

dlmarion commented on core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:
```diff
@@ -611,6 +611,12 @@ void setProperty(String tableName, String property, String value)
    * properties but has the same behavior and better documentation.
    *
+   * <p>
+   * Accumulo has multiple layers of properties that for many APIs and SPIs presented as single
```
same comment

EdColeman commented:
I created https://github.com/apache/accumulo/issues/2996 to track things specifically to the CMEs that are occurring.  Depending on the final state of this PR, that issue can be closed - or can be used for follow-on work independent of the PR so this PR can proceed without those issues getting lost

dlmarion commented:
Full IT build passed successfully

dlmarion reviewed (approved):
LGTM, I think the log messages should be modified to note that we are going to retry. In the case where it's continually failing it's going to be obvious that it's retrying. In the case where it fails once and then succeeds, you won't know by looking at the log that it was successful. To that end, it may be useful to log that it was successful when a retry was performed.

keith-turner commented:
> LGTM, I think the log messages should be modified to note that we are going to retry. 

The Retry code will append the following to the log message to indicate that its going to retry.

https://github.com/apache/accumulo/blob/6976933640be998c01c2a47c4a1bf8d0dca84d86/core/src/main/java/org/apache/accumulo/fate/util/Retry.java#L239-L242

> To that end, it may be useful to log that it was successful when a retry was performed.

That would be really nice to do.  Looking at the Retry class it does not support this.  I am going to look into doing this and see if it should be done in this PR or as a follow on PR that makes a more general change to Retry.

keith-turner commented:
> To that end, it may be useful to log that it was successful when a retry was performed.

@dlmarion I added logging of when a an operation completed in 0c3bd62.  I also opened #3007 with something else i noticed and that issue has output from the new logging.

dlmarion commented:
Looks good to me.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java:
```diff
@@ -119,6 +127,38 @@ public void modifyProperties(final Consumer<Map<String,String>> mapMutator)
+          throw new RuntimeException(e);
+        }
+      } finally {
+        retry.useRetry();
```
I wonder if instead of calling the method `Retry.logCompletion`, it should be called `Retry.completed(msg)` and then it disables any checks that might throw errors. For example:

I don't think this is a problem here, because this Retry is configured for infiniteRetries. However, if Retry.useRetry is called elsewhere in a finally block, it could throw an exception even if the operation completed successfully (in the case where the Retry was configured for N retries, and it had retried N times but succeeding on the last one.

dlmarion replied:
Looking further, it looks like the issue can be avoided, if in a finally block (and maybe elsewhere where useRetry is called:

```
   ...
   } finally {
     if (retry.canRetry()) {
      retry.useRetry();
     }
  }
```

keith-turner replied:
Using the Rety class I find myself wondering if it can be improved and then more widely used in the code. Like maybe pass an interface to a method that will do the retries, looping, logging etc.  The interface can supply the operation to run and the messages to log.

keith-turner replied:
Continuation or prev comment.  Wondering about something like the following.  Thinking through it a bit i think dealing with checked exceptions would be the biggest pain point in doing something like this.

```java
Retry retry = ..
retry.execute(new Retriable(){
    //whatever is needed to execute a retriable operation : Runnable, descriptive message, etc
});
```

dlmarion replied:
I did something similar when we were working on external compactions to retry thift calls - [RetryableThriftCall](https://github.com/apache/accumulo/blob/main/server/base/src/main/java/org/apache/accumulo/server/compaction/RetryableThriftCall.java)

keith-turner replied:
> I did something similar when we were working on external compactions to retry thift calls - [RetryableThriftCall](https://github.com/apache/accumulo/blob/main/server/base/src/main/java/org/apache/accumulo/server/compaction/RetryableThriftCall.java)

We need something more general like that so we can reuse retry code.  Still scratching my head over the best way to handle checked exceptions though.  Like what the most concise way to handle one operation that throws checked exception X and another that throws checked exception Y and these are checked exceptions that we want to pass up.

dlmarion replied:
I think this is a case that's similar to `Future.get()` where it throws an `ExecutionException`. As an example, we could:

1. modify  [RetryableThriftFunction](https://github.com/apache/accumulo/blob/main/server/base/src/main/java/org/apache/accumulo/server/compaction/RetryableThriftFunction.java) to declare `Exception` rather than `TException'. 
2. Then in [RetryableThrifCall](https://github.com/apache/accumulo/blob/main/server/base/src/main/java/org/apache/accumulo/server/compaction/RetryableThriftCall.java#L106) catch `Exception`
3. When retries have been exhausted, throw a RetriesExceededException [here](https://github.com/apache/accumulo/blob/main/server/base/src/main/java/org/apache/accumulo/server/compaction/RetryableThriftCall.java#L118) when there is no error and throw an ExecutionException when retries have been exhausted and there is an error.

keith-turner replied:
@dlmarion I think it would be good to open up some sort of follow on issue about improving retry, but not sure what exactly.  I think you have more defined thoughts on it.

cshannon commented:
I've been able to pretty consistently in my IDE get the `PermissionsIT.systemPermissionsTest()` to fail with a CME. It's not every time but often. I ran the test against this branch and can confirm that this reliably fixes the test by doing a retry. Here is the output from the logs.

`2022-10-07T10:49:48,842 [admin.TableOperations] DEBUG: Unable to modify table properties for PermissionsIT_systemPermissionsTest0__ALTER_TABLE_WITHOUT_PERM_TEST__ because of concurrent modification, retrying attempt 1 (suppressing retry messages for 180000ms)
2022-10-07T10:49:48,843 [util.Retry] DEBUG: Sleeping for 25ms before retrying operation
2022-10-07T10:49:48,876 [admin.TableOperations] DEBUG: Modifying properties for table PermissionsIT_systemPermissionsTest0__ALTER_TABLE_WITHOUT_PERM_TEST__ completed after 2 retries and is no longer retrying.
2022-10-07T10:49:48,969 [functional.PermissionsIT] DEBUG: Confirming that the presence of the ALTER_TABLE permission properly permits the user`

dlmarion commented:
@keith-turner  - thoughts on merging this? Are you waiting on anything in particular?

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java:
```diff
@@ -57,19 +56,86 @@ void setProperty(final String property, final String value)
    */
-  void modifyProperties(Consumer<Map<String,String>> mapMutator) throws AccumuloException,
-      AccumuloSecurityException, IllegalArgumentException, ConcurrentModificationException;
+  Map<String,String> modifyProperties(Consumer<Map<String,String>> mapMutator)
```
I think `computeProperties` might also be a decent name, since it is now returning the properties. This is analogous to the `compute*()` methods that return the resulting value in the Java Map APIs.

keith-turner replied:
I think i like `modify` a bit better because it conveys something that is happening in the distributed system where `compute` feels a bit more centered on what is happening in the client.

ctubbsii reviewed (approved):
Minor name suggestion, if you like it better, but does not matter much to me. The current name is also fine.
Have not manually tested, but am okay with this if it follows the behavior that we discussed.
